transformations/change_char_case/transformation.py

```diff
@@ -3,9 +3,7 @@
 from interfaces.SentenceOperation import SentenceOperation
 from tasks.TaskTypes import TaskType
 
-
-def change_char_case(text, prob=0.1, seed=0, max_outputs=1):
-    random.seed(seed)
+def change_char_case(text, prob=0.1, max_outputs=1):
```

Comment:
Hi @SirRob1997 : I went through the build log. The pytest is failing for only ChangeCharCase transformation. The reason is that this transformation uses random function in the logic and a seed was used in the earlier version of the code. 

In the current logic, seed is global and this random function is not seeded. If you fix this, probably the build will go through successfully (or we may encounter more bugs).

Reply:
Hi @AbinayaM02, thanks for taking a look at this! This was a common pattern among many transformations, hence this PR. The random function should in fact be seeded, afaik the global seed gets set for the tests [here](https://github.com/GEM-benchmark/NL-Augmenter/blob/b221b970f946a376d6db867e1c7e143376c1b994/test/test_main.py#L106) and seeding is shared among multiple `import random` in different files. Running the tests also creates reproducible results on local which verifies that `random` is in fact already seeded properly. 

I believe there is a problem with the GitHub action that evaluates the new code against the already existing tests instead of the slightly adapted ones provided in this PR or am I missing something here?

Reply:
This is from the Github Actions log,

AssertionError: Mis-match in expected and predicted output for ChangeCharCase transformation: 
E              Expected Output: Andrew finally returned the French book tO ChriS that I bought Last week. 
E              Predicted Output: AndreW finally returned The French booK To ChRis that I bougHt last week.

I checked the test.json in the PR for this ChangeCharCase transformation. The expected output in the test.json is picked correctly (first line) but the actual execution of the transformation during pytest is producing a different output (second line). So, I thought this has to do with the random.random() call in the code.

Reply:
Hmm, yeah you're right that is weird. Any idea why this is happening in the Github action and not on local?

Reply:
I'm not completely sure about why this is happening. Will dig a bit deeper.

There is a strange pattern in all the test.json files. Only 2 or 3 out of 5 sample outputs are changed by the global seed changes. I was expecting all the sentence outputs to have some changes since the new global seed might be different from what was used earlier in the individual modules. Either there should be no changes at all in the test.json files [if the seeds before and after the current PR are same] else all the output samples should have change. 

Is my understanding right?

Reply:
Yeah that's right, I noticed that too and opened up issue #162 but I brushed it off as random chance. I haven't found a reason for why that pattern would occur. If I recall correctly, for some of the transformations there were changes in the `test.json` needed even though the `GLOBAL_SEED` matched the previously used one which I found rather odd. However, I couldn't find a root cause for it. I was thinking that I missed a place where seeding was additionally done but a project wide search for "seed" seemed to reject that hypothesis.

Reply:
Let me try to reproduce your fix from my end and see if it happened by chance or is it some bug. Will update soon!

Reply:
Hey @SirRob1997 : I'm getting the exact error on the Github actions log on my local as well. Could you please execute the following command and let me know if you also get the same error or it passes without any issue.

```python
pytest -v --t=light
```

Reply:
Yes, that indeed fails for more as well. However, this still passes all the tests for this transformation:
```
pytest -s --t=change_char_case
```

Reply:
Yes, it failed for me too. 

I think I found the issue after breaking my head for so long! The issue is with the random.random() call in the change_char_case transformation.

 I did the following to narrow down the issue,
1. Created new test.json file for the change_char_case transformation after leveraging the global seed [by calling the initialize_model() method only for test file creation].
2. Ran the pytest for change_char_case transformation and all light transformations. In both the cases, it failed for change_char_case transformation.
3. Tried printing the random.random() call. My assumption was that it should use the global seed and give me the same random numbers both during pytest check and while running transforamtion.py separately but it didn't.
4. Finally, I realized that only test case generation is leveraging the global seed but the actual transformation is not using it. So, called set_seed() function in the  change_char_case method. 

```python
from initialize import set_seed

def change_char_case(text, prob=0.1, max_outputs=1):
    set_seed()
    results = []
    for _ in range(max_outputs):
        result = []
        for c in text:
            x = random.random()
            y = random.random()
            print(x, y)
            if c.isupper() and x < prob:
                result.append(c.lower())
            elif c.islower() and y < prob:
                result.append(c.upper())
            else:
                result.append(c)
        result = "".join(result)
        results.append(result)
    return results
```

Once, I made the above change, the testcases for change_char_case passed (both individually and as part of the light transformations). And I landed up in another test case failure for ChangePersonNamedEntities transformation!

So, to keep the global seed changes everywhere, I think you need to check which all transformations use seed, and call set_seed() before performing transformation. Also the test.json files should be recreated for all of them.

Reply:
Yeah, good catch for sure. It's still weird to me that `set_seed()` in the `test_main.py` doesn't handle this properly since I believe it should. I wouldn't want to add  `set_seed()` to every transformation since that seems very unclean... Would be nice to figure out why exactly this isn't properly handled from the `test_main.py`.